torch/distributed/distributed_c10d.py

```diff
@@ -761,12 +761,26 @@ def _new_process_group_helper(
             assert backend.upper() in Backend._plugins, (
                 f"unknown c10d backend type {backend.upper()}"
             )
-            pg = Backend._plugins[backend.upper()](
-                prefix_store, rank, world_size, timeout
+
+            backend_creator_fn = Backend._plugins[backend.upper()]
+
+            if pg_options is None:
+                import inspect
+                backend_module = inspect.getmodule(backend_creator_fn)
+                assert 'Options' in dict(inspect.getmembers(backend_module, inspect.isclass)), (
+                    f"Process group Options not defined in {backend.upper()} implementation"
+                )
+
+                pg_options = backend_module.Options.create()
+
+            pg = backend_creator_fn(
+                _group_count, prefix_store, group_rank, group_size, global_ranks_in_group, pg_options, timeout
```

Comment:
I guess adding `global_ranks_in_group` and `pg_options` might be fine here, but it doesn't seem like passing in `_group_count` is a good idea. This is an internal implementation details that we should not expose.

There also might be a BC issue here where we would break existing `backend_creator_fns` with this change.

Reply:
The `_group_count` is used as a group identifier. When a collective is called on a specific process group object, for 3rd party backends there's currently no support to associate from what group is the call coming from. Including a group identifier bridges this gap.

Agree with BC issue, and happy to hear suggestions here. `global_ranks_in_group` and `pg_options` are currently being not passed to 3rd party, which is the main limitation trying to address in this PR.

Reply:
I want to check back on this point on the point about backward compatibility here. We can incorporate an API version number in 3rd party distributed backends and based on that invoke the existing signature or the new one proposed. Does that sound good?

Reply:
We should reorder the arguments so the new ones are at the end. 
Instead of breaking existing backends, we should check the number of arguments the creation function takes - and emit a warning if too few.

Reply:
@kumpera Alternatively, does extending `dist.register_backend()` to take in an API version (string) info sound reasonable to support backward compatibility? This would be an optional argument, so during dispatching if version is omitted or say v1.0 we can send only the limited number of arguments as done today.

Reply:
I want to get some thoughts to unblock this PR. The main issue is the inability to inspect the function signature at runtime for the backend creator function. Python inspect module fails for the same reason as in this open issue [28233](https://github.com/pytorch/pytorch/issues/28233). Also, the definition and python binding of this function are in the control of the 3rd party provider.

So in this regard, what about specifically asking the 3rd party provider to specify the API version. For existing backends we can assume the API to be v1.0, which accepts the limited set of arguments.

If this sounds good, I can provide the implementation in the next iteration. Let me know.

Reply:
That sounds the path forward. Do you think something along the lines of:
```
class BackendProtocol(Enum):
 ONE = auto() # original 4 args func
 TWO = auto() # new version with 6 args
class Backend:
    def register_backend(cls, name, func, version=BackendProtocol.One):
      ....
```

Reply:
Yes, this is along the lines of what I meant. 
I also have another approach that might be little extensible than this. That would be to define an overload `register_backend` function that'll accept a function that takes a predefined struct instead of individual arguments. The struct can be defined like how ProcessGroupOptions or Store classes are defined. 

I'll make an update, so you can take a pass.

Reply:
I've added a flag, `extended_api` to differentiate between old and new APIs when registering 3P communication backends. When using the `extended_api=True` the registered backend is expected to accept a new `c10d::DistributedBackendOptions` instance along with any process group options implemented by that backend extending the `c10d::ProcessGroup::Options`. 

The `c10d::DistributedBackendOptions` gives the ability to add future changes to the backend registration contract without breaking backward compatibility.